### ISSUE TITLE
First incoming HTTP connection is delayed by 300ms on Windows

### ISSUE THREAD

**Issue by mitranim** [denoland/deno#10762]
## Environment

```
deno 1.10.1 (release, x86_64-pc-windows-msvc)
v8 9.1.269.27
typescript 4.2.2
```

The problem is specific to the Windows version of Deno. It does _not_ happen for `x86_64-unknown-linux-gnu` under WSL2, but it _does_ happen for `x86_64-pc-windows-msvc` invoked under WSL2.

## Problem

On Windows, after starting a Deno HTTP server (either stdlib between `0.60.0` and `0.97.0` or `Deno.listen`), the first incoming connection is delayed by 300ms.

When doing rapid development in watch-and-restart mode, this can be a minor annoyance.

Repro scripts:

* [deno_srv_delay_stdlib.js](https://gist.github.com/mitranim/ea15f621505d38877190a0a057b5cd01)
* [deno_srv_delay_unstable.js](https://gist.github.com/mitranim/e35e95e0d609e3167effaca2232c4639)

Typical output from repro scripts (`x86_64-pc-windows-msvc`):

```
[srv] listening on http://localhost:57294
headers: 316ms
body: 1ms
headers: 1ms
body: 0ms
```

Typical output from repro scripts (`x86_64-unknown-linux-gnu`):

```
[srv] listening on http://localhost:57294
headers: 7ms
body: 1ms
headers: 0ms
body: 0ms
```

**Comment by MarkTiedemann** [comment]
The delay is from trying to connect via IPv6, failing, then switching to IPv4.

`test.ts`
```typescript
for await (let conn of Deno.listen({ port: 80 })) {
  for await (let event of Deno.serveHttp(conn)) {
    event.respondWith(new Response(Deno.pid.toString()));
  }
}
```
`test.cmd`
```batch
@echo off
start "" /b deno run --allow-net --unstable test.ts
timeit curl -s 127.0.0.1 2>&1 | findstr Elapsed
timeit curl --ipv4 -s localhost 2>&1 | findstr Elapsed
timeit curl -s localhost 2>&1 | findstr Elapsed
curl --ipv6 -sS localhost
for /f %%p in ('curl -s localhost') do taskkill /pid %%p /f > nul
```

```
C:\> test
Elapsed Time:     0:00:00.030
Elapsed Time:     0:00:00.080
Elapsed Time:     0:00:00.244
curl: (7) Failed to connect to localhost port 80: Connection refused
```

**Comment by mitranim** [comment]
Nice find. For some reason, explicitly listening on `localhost` seems to eliminate the delay:

```js
import * as hs from 'https://deno.land/std@0.97.0/http/server.ts'
const srv = hs.serve({port, hostname: 'localhost'})
```

```js
const srv = Deno.listen({port, hostname: 'localhost'})
```

Works for local development servers. Would be nice if the workaround wasn't necessary.

**Comment by kitsonk** [comment]
Yes, that is because it binds to whatever the `localhost` host name resolves to, which would be the IPv6 address, where as the default if emitted is `127.0.0.1` which is the IPv4 address, therefore connecting on IPv6 initially, which it isn't bound to, is what is occurring.

@bnoordhuis @piscisaureus any ideas if we changed our default to `localhost` or somehow detected IPv6 or something would resolve the problem?

**Comment by lucacasonato** [comment]
We might want to align to node behaviour here:

> If host is omitted, the server will accept connections on the unspecified IPv6 address (::) when IPv6 is available, or the unspecified IPv4 address (0.0.0.0) otherwise.

> In most operating systems, listening to the unspecified IPv6 address (::) may cause the net.Server to also listen on the unspecified IPv4 address (0.0.0.0).

**Comment by MarkTiedemann** [comment]
> In most operating systems, listening to the unspecified IPv6 address (::) may cause the net.Server to also listen on the unspecified IPv4 address (0.0.0.0).

That's different from the current behavior.

When specifying `Deno.listen({ port: 80, hostname: "[::]" })`, `curl [::1]` runs fine, whereas `curl 127.0.0.1` shows a `Connection refused` error.

**Comment by lucacasonato** [comment]
Yeah - I am also not super fond of the second behaviour, unless we introduce a `tcp4` / `tcp6` "protocol" next to the existing `tcp` protocol.

**Comment by kitsonk** [comment]
That just makes it more difficult for people to figure out, when an explicit IP v4 or v6 address resolves the problem and if connections are refused on `127.0.0.1` we document that they need to be explicit.  Setting `tcp4` or `tcp6` is effectively the same as being explicit about the host of `127.0.0.1` or `[::]`.

**Comment by MarkTiedemann** [comment]
I find explicitly listening to both IPv4 and IPv6 a bit cumbersome:

```typescript
async function handle(listener: Deno.Listener) {
  for await (let conn of listener) {
    for await (let event of Deno.serveHttp(conn)) {
      event.respondWith(new Response());
    }
  }
}

handle(Deno.listen({ port: 80, hostname: "0.0.0.0" }));
handle(Deno.listen({ port: 80, hostname: "[::]" }));
```

If it was possible to specify a `string[]` for the `hostname` option, it could default to `["0.0.0.0", "[::]"]` (rather than just `"0.0.0.0"`).

**Comment by lucacasonato** [comment]
I dislike being able to specify multiple hostnames. We would have to multiplex listeners under the hood then.

For reference this is how it works in Go:

> Listen announces on the local network address.
>
> The network must be "tcp", "tcp4", "tcp6", "unix" or "unixpacket".
>
> For TCP networks, if the host in the address parameter is empty or a literal unspecified IP address, Listen listens on all available unicast and anycast IP addresses of the local system. To only use IPv4, use network "tcp4". The address can use a host name, but this is not recommended, because it will create a listener for at most one of the host's IP addresses. If the port in the address parameter is empty or "0", as in "127.0.0.1:" or "[::1]:0", a port number is automatically chosen. The Addr method of Listener can be used to discover the chosen port.

I think instead we should change the default value of `hostname` to the empty string, and then have empty string mean "all available unicast and anycast IP addresses of the local system", just like in Go. `tcp4` and `tcp6` transport types _could_ then be introduced later to mean only IPv4 addresses, or only IPv6 addresses.

**Comment by piscisaureus** [comment]
Maybe it would be an idea to support dual stack sockets?

This is controlled by the `IPV6_V6ONLY` socket option; I believe it's off by default on linux (which means a dual-stack is enabled), but on windows it is [on by default (dual-stack disabled)](https://docs.microsoft.com/en-us/windows/win32/winsock/dual-stack-sockets).

**Comment by mitranim** [comment]
Perhaps pertinent: in Node (v16), the problem appears to be reproducible in the exact opposite way:

```js
import * as ht from 'http'

const srv = new ht.Server()

// Doesn't seem to cause first connection delay on Windows.
srv.listen(somePort)

// Seems to cause first connection delay on Windows (sometimes?).
srv.listen(somePort, 'localhost')
```

**Comment by masx200** [comment]
> The delay is from trying to connect via IPv6, failing, then switching to IPv4.

For this problem, it is perfectly possible to use `Promise.any` to connect ipv6 and ipv4 at the same time. This should cause no delay.
